### PR TITLE
feat(centraldashboard): Manage multiple profiles

### PR DESCRIPTION
**Pull request by frazs**
**Incomplete WIP, do not merge:** to be continued by somebody else, as I am switching over to another team.

Will resolve https://github.com/StatCan/kubeflow/issues/7 when finished.

Preliminary steps to work on this:
1. Clone this branch and navigate to `kubeflow/components/centraldashboard`
2. Install node **v12.18.3** and make sure you are using it (`node -v`). [nvm](https://github.com/nvm-sh/nvm) is very handy for this!
3. `make build-local`
4. `kubectl apply` another [profile](https://www.kubeflow.org/docs/components/multi-tenancy/getting-started/#manual-profile-creation) setting yourself as the owner. No need for the `resourceQuotaSpec` section, that can be removed entirely. If you lack permissions for this, ping me and I can do that for you.
5. In another terminal, on a cluster running Kubeflow, do `kubectl port-forward -n kubeflow deployment/profiles-deployment 8081:8081`. This will enable the Manage Contributors section. Other areas of the sidebar and main page will remain unconnected in the following step, but they are not necessary for this issue.
6. Back in the centraldashboard directory, run `USERID_HEADER=kubeflow-userid USERID_PREFIX= KF_USER_ID=[name@cloud.statcan.ca, e.g. frances.zsurka@cloud.statcan.ca] npm run dev`. 

Sometimes for me the last step will not run, complaining of too many watchers. This seems to be a RAM issue on my end (or at least I've tried increasing `fs.inotify.max_user_watches` without effect). Closing my browser and/or VS Code helps. Re-opening them after the server has started works fine, and the server automatically applies code changes. 

What I got working:
- In `main-page.js`, when the array of namespaces for a user is loaded, it is now sorted into an array of "owner" role namespaces and other role namespaces. The owner array is put before the other array, so that owned namespaces will appear at the top of the namespace dropdown as well as the namespace membership list on the Manage Contributors page. 
- The owned namespaces array is set as the value of a new main page class property, `this.multiOwnedNamespaces`. multiOwnedNamespaces is something I would like to have a better name for later, as it is confusing (these are multiple owned namespaces, not e.g. namespaces owned by multiple users), but for now I wanted something visually distinct from the existing singular property `ownedNamespace`. If all instances of `ownedNamespace` get refactored away, we could change `multiOwnedNamespaces` to the more intuitive `ownedNamespaces`.
- In `main-page.pug`, `multi-owned-namespaces='[[multiOwnedNamespaces]]'` is passed to manage-users-view. This specific case combination, `dash-case=[[camelCase]]`, has to be used for this to work. Not sure if that's a quirk of pug, Polymer, or something else.
- I reworked the namespace membership list in Manage Contributors. This was originally just an experiment to ensure that things were getting passed properly, but I found the result more usable. Originally, the membership list only ever has two rows. The alphabetically first owned namespace is "Owner", and every other namespace on a single line is "Contributor". If there are several other namespaces, at some point they are cut off with a "...". The new membership list has a row per namespace, lists the owner roles first, and hypothetically accepts roles other than "Contributor" (not that any exist now). A helper function capitalizes the first letter of the role name to be in line with the original appearance.
 - All owned namespaces, and only owned namespaces, are listed in multiple "Contributors to your namespace - [namespace name]" section, with the appropriate namespace name and the appropriate list of contributors. I had considered an alternate approach where a single owned namespace would be selected from the dropdown, which would set and continue to use the singular ownedNamespace. But that presented its own challenges and seems a lot clunkier: I think it makes the most sense to see the contributors to your namespaces, and thus be able to tell exactly where a particular contributor is/isn't, all at once. But if the issues below can't be easily resolved, perhaps the one-at-a-time approach is worth reconsidering. 

Where I got stuck:
The list of multiple owned namespaces was done with a `dom-repeat`, but that may be the wrong method (or something may be missing from it). When a new email is typed into any contributor field (`Add by email address`), character by character, it appears in _every_ contributor field. I'm still not entirely sure what's going on under the hood here or how to properly instantiate/differentiate these elements. It might have something to do with the `this.contributorInputEl = this.$.ContribEmail;` line in `manage-users-view.js`.

Furthermore, I could not get the add or remove contributor iron-ajax methods to work. In this commit, I left them in their default state, where any email entered, or removal enacted, will apply to only the alphabetically first owned namespace. There is no update for this visually until a refresh. If I move them to the same location I moved `GetContribsAjax` to, and have them likewise target `[[multiOwnedNamespace.namespace]]`, nothing happens: console log will say that `api` is undefined, with the corresponding failing code in `main-users-view.js` being `const api = this.$.AddContribAjax;` and `const api = this.$.RemoveContribAjax;`. 

So if I'm even on the right track, it may be important to puzzle out exactly how `this.$` behaves and how that interacts with `dom-repeat`. Some old discussions and documentation I've come across note that `this.$` cannot get elements from inside of a `dom-repeat`, but they seem to be for older versions of Polymer, so I'm not certain whether that's still the case (or what the best workaround is).

I've also tried the iron-ajax lower down where the contributors are being templated, as well as its `model` syntax. No luck, but might not have found the appropriate combination if one exists.

Note that any Polymer syntax found in documentation and discussions will look different from what's in the .pug files. Pug is a templating engine that will compile into the appropriate code.

What remains:
- [X] Successfully add contributors to a specific owned namespace
- [x] Successfully remove contributors from a specific owned namespace
- [x] Confirm whether there are more AJAX requests than necessary (and if so refactor)
- [x] Check if there are any other problematic areas where a singular `ownedNamespace` is assumed
-  If it is possible to refactor all singular instances of `ownedNamespace` into plural ones, consider renaming what is currently `multiOwnedNamespaces` to just `ownedNamespaces`
- [x] If any of the not-directly-related changes (dropdown re-order, Namespace memberships list change) are deemed unnecessary or suited for other issues, revert them (and if necessary replace with a different way to get owned namespaces)
- [x] If necessary, update `public/components/main-page_test.js` [(Jasmine)](https://jasmine.github.io/) so that the docker image can build.
- [x] Update `public/components/main-users-view_test.js` [(Jasmine)](https://jasmine.github.io/) so that the docker image can build.

A side-issue:

I noticed that when after being the owner of more than one profile, when I refresh my Notebook Servers page (on the regular production website; haven't gotten this to load with `npm run dev`), sometimes but not always, it will switch to the first owned namespace alphabetically. 

Whether the switch happens seems to depend on whether my URL specifies e.g. `_/jupyter/?ns=frances-zsurka` (no switch) or not (just `_/jupyter/` switches on refresh). And if I have the `?ns=frances-zsurka` parameter, and I refresh that page, then click anywhere on that page, that `?ns=frances-zsurka` parameter is lost and the following refresh will switch to the first owned namespace alphabetically. 

It is a minor annoyance and most likely suited to a separate issue (I suggest creating one, perhaps after this is sorted and multiple namespaces will be more broadly used), but I am mentioning it here in case there's a way to refactor to multiple owned namespaces that would also address this. If not another avenue would be to prevent the loss of `?ns=[namespace]` parameter.

### CODE REVIEW

**Comment by wg102** [comment]
This is ready to review, but will need to be merged with stc-master after the failing tests are fixed.

**Comment by wg102** [comment]
Code was rebased of stc-master with the fixes. Ready for review. 
The one task that is not done is the refactor, I found it confusing to read in the code if the only difference is an 's'.

File: components/centraldashboard/public/components/manage-users-view.js [hunk 1]
```diff
@@ -35,55 +34,43 @@ export class ManageUsersView extends utilitiesMixin(PolymerElement) {
             isClusterAdmin: {type: Boolean, value: false},
             namespaces: Array,
             ownedNamespace: {type: Object, value: () => ({})},
-            newContribEmail: String,
-            contribError: Object,
-            contributorInputEl: Object,
```
Comment (brendangadd) [inline components/centraldashboard/public/components/manage-users-view.js:40]: Also add your new property.

File: components/centraldashboard/public/components/manage-users-view.js [hunk 2]
```diff
@@ -35,55 +34,43 @@ export class ManageUsersView extends utilitiesMixin(PolymerElement) {
             isClusterAdmin: {type: Boolean, value: false},
             namespaces: Array,
             ownedNamespace: {type: Object, value: () => ({})},
```
Comment (brendangadd) [inline components/centraldashboard/public/components/manage-users-view.js:36]: Looks like `ownedNamespace` is no longer used/applicable.

**Review by brendangadd — CHANGES_REQUESTED** [review]

**Review by brendangadd — APPROVED** [review]
Looks great! I'll merge later to remind myself to deploy later when we have time to do some extra QA once it hits the cluster.

**Pull request merged**